Spring.FluentContext/ContextRegistrations.cs

```diff
@@ -0,0 +1,210 @@
+		{
+			return RegisterNamedProxyFactory<T>(ctx, IdGenerator<T>.GetUniqueId());
+		}
+
+		/// <summary>
+		/// Registers singleton <c>instance</c> with specified <c>id</c>.
+		/// </summary>
+		/// <typeparam name="T">Type of singleton instance.</typeparam>
+		/// <param name="id">Id of registered object.</param>
+		/// <param name="instance">Singleton instance to register.</param>
+		/// <returns>Registered object reference.</returns>
+		public static IObjectRef<T> RegisterNamedSingleton<T>(this IConfigurableApplicationContext ctx, string id, T instance)
+		{
+			SafeGetObjectFactory(ctx).RegisterSingleton(id, instance);
+			return new ObjectRef<T>(id);
+		}
+
+		/// <summary>
+		/// Registers singleton <c>instance</c> with default id.
+		/// </summary>
+		/// <typeparam name="T">Type of singleton instance.</typeparam>		
+		/// <param name="instance">Singleton instance to register.</param>
+		/// <returns>Registered object reference.</returns>
+		public static IObjectRef<T> RegisterDefaultSingleton<T>(this IConfigurableApplicationContext ctx, T instance)
+		{
+			return RegisterNamedSingleton(ctx, IdGenerator<T>.GetDefaultId(), instance);
+		}
+
+		/// <summary>
+		/// Registers singleton <c>instance</c> with unique id.
+		/// </summary>
+		/// <typeparam name="T">Type of singleton instance.</typeparam>		
+		/// <param name="instance">Singleton instance to register.</param>
+		/// <returns>Registered object reference.</returns>
+		public static IObjectRef<T> RegisterUniquelyNamedSingleton<T>(this IConfigurableApplicationContext ctx, T instance)
+		{
+			return RegisterNamedSingleton(ctx, IdGenerator<T>.GetUniqueId(), instance);
+		}
+		
+		/// <summary>
+		/// Given a object name, create an alias. We typically use this method to
+		/// support names that are illegal within XML ids (used for object names).
+		/// </summary>
+		/// <param name="name">The name of the object.</param>
+		/// <param name="theAlias">The alias that will behave the same as the object name.</param>
+		/// <exception cref="Spring.Objects.Factory.NoSuchObjectDefinitionException">
+		/// If there is no object with the given name.
+		/// </exception>
+		/// <exception cref="Spring.Objects.Factory.ObjectDefinitionStoreException">
+		/// If the alias is already in use.
+		/// </exception>
+		public static void RegisterAlias(this IConfigurableApplicationContext ctx, string name, string theAlias)
+		{
+			SafeGetObjectFactory(ctx).RegisterAlias(name, theAlias);
+		}
+		
+		/// <summary>
+		/// Protects access to the internal object factory used by the ApplicationContext if attempted to be accessed when in improper state.
+		/// </summary>
+		/// <returns>The internal ObjectFactory used by the ApplicationContext</returns>
+		/// <exception cref="System.InvalidOperationException">Cannot Access ApplicationContext in this state!</exception>
+		private static IConfigurableListableObjectFactory SafeGetObjectFactory(IConfigurableApplicationContext ctx)
```

Comment:
This method does not looks good :(

* it violates [Open-Closed Principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) by casting interface to specific implementations,
* it uses reflection to access private member of `AbstractApplicationContext`,
* it uses the `TemporaryInitObjectFactory` property of the interface where it's name tells there is something wrong about it ;-) It almost looks like this should be some internal, maybe temporary state, but it's exposed to the public.

Can you think of the better implementation of this problem? Unfortunately, you did not provided any sample usages of those new fluent contexts, so I cannot check what problems you are exactly solving here and what would be the usability of those members. **Please add examples**, as it will help manage the project but also will help others to understand it.


Briefly, one alternative solution that comes to my mind is to make an interface exposing the `ObjectFactory` and implement it on all fluent contexts. You already have those implementations in place so it won't be hard.

The bigger question here I have is why you cannot just use the `ctx.ObjectFactory` as it is obviously there? Again, example would help.

Reply:
This whole stuff tried to solve injecting Fluent config into `AbstractXmlApplicationContext` line 149 without the need of duplicating too many code from existing classes. I might found a better solution now, will give a try, please hold